File: articles/gradient-descent-and-optimization-in-machine-learning/fortran/gradient_descent.f90
program gradient_descent
  implicit none
  real :: x(5) = (/ -2.0, -1.0, 0.0, 1.0, 2.0 /)
  real :: y(5) = (/ -2.85, -0.67, 1.47, 3.63, 5.82 /)
  real :: w, b, eta, grad_w, grad_b, err, loss
  integer :: i, step

  w = 0.0
  b = 0.0
  eta = 0.08

  do step = 1, 80
    grad_w = 0.0
    grad_b = 0.0
    do i = 1, 5
      err = (w*x(i) + b) - y(i)
      grad_w = grad_w + (2.0/5.0) * err * x(i)
      grad_b = grad_b + (2.0/5.0) * err
    end do
    w = w - eta * grad_w
    b = b - eta * grad_b
  end do

  loss = 0.0
  do i = 1, 5
    loss = loss + (y(i) - (w*x(i)+b))**2
  end do
  loss = loss / 5.0

  print *, 'weight=', w, ' bias=', b, ' loss=', loss
end program gradient_descent

File: articles/gradient-descent-and-optimization-in-machine-learning/typescript/gradient_descent.ts
type Point = { x: number; y: number };

const rows: Point[] = [
  { x: -2, y: -2.85 },
  { x: -1, y: -0.67 },
  { x: 0, y: 1.47 },
  { x: 1, y: 3.63 },
  { x: 2, y: 5.82 },
];

function mse(weight: number, bias: number): number {
  return rows.reduce((sum, row) => sum + Math.pow(row.y - (weight * row.x + bias), 2), 0) / rows.length;
}

let weight = 0;
let bias = 0;
const eta = 0.08;

for (let step = 0; step < 80; step += 1) {
  let gradW = 0;
  let gradB = 0;
  for (const row of rows) {
    const err = (weight * row.x + bias) - row.y;
    gradW += (2 / rows.length) * err * row.x;
    gradB += (2 / rows.length) * err;
  }
  weight -= eta * gradW;
  bias -= eta * gradB;
}

console.log({ weight, bias, loss: mse(weight, bias) });
